src/webnn/native/dml/DMLUtils.cpp

```diff
@@ -52,42 +61,396 @@ namespace webnn::native::dml {
-                               ComPtr<ID3D12Device> D3D12Device) {
-        WEBNN_CHECK(commandList->Close());
-        ID3D12CommandList* commandLists[] = {commandList.Get()};
-        commandQueue->ExecuteCommandLists(ARRAYSIZE(commandLists), commandLists);
-        WEBNN_CHECK(commandQueue.Get()->GetDevice(IID_PPV_ARGS(D3D12Device.GetAddressOf())));
-        ComPtr<ID3D12Fence> fence;
+    void Device::FillUploadResourceAndInputBindings(
+        uint64_t uploadResourceSize,
+        std::vector<DML_BUFFER_BINDING>& inputBufferBinding,
+        const std::vector<std::shared_ptr<InputNode>>& inputNodes,
+        std::unordered_map<std::string, Input> namedInputs) {
+        D3D12_RANGE uploadBufferRange{0, uploadResourceSize};
+        int8_t* uploadBuffer;
         WEBNN_CHECK(
-            D3D12Device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(fence.GetAddressOf())));
-        WEBNN_CHECK(commandQueue.Get()->Signal(fence.Get(), 1));
-        WEBNN_CHECK(fence->SetEventOnCompletion(1, nullptr));
-        WEBNN_CHECK(commandAllocator->Reset());
-        WEBNN_CHECK(commandList->Reset(commandAllocator.Get(), nullptr));
+            mUploadResource->Map(0, &uploadBufferRange, reinterpret_cast<void**>(&uploadBuffer)));
+        uint64_t offset = 0;
+        for (size_t i = 0; i < inputNodes.size(); ++i) {
+            auto inputNode = inputNodes[i];
+            if (namedInputs.empty()) {
+                if (inputNode->type == NodeType::ConstantInput) {
+                    offset =
+                        RoundUpToMultiple(offset, (uint64_t)DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT);
+                    inputBufferBinding[i].Buffer = mInputResource.Get();
+                    inputBufferBinding[i].Offset = offset;
+                    inputBufferBinding[i].SizeInBytes = inputNode->byteLength;
+                    memcpy(uploadBuffer + offset, inputNode->buffer,
+                           static_cast<size_t>(inputNode->byteLength));
+                    offset = offset + inputNode->byteLength;
+                }
+            } else {
+                if (inputNode->type == NodeType::NonConstantInput) {
+                    offset =
+                        RoundUpToMultiple(offset, (uint64_t)DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT);
+                    auto arrayBufferView = namedInputs[inputNode->name].resource.arrayBufferView;
+                    inputBufferBinding[i].Buffer = mInputResource.Get();
+                    inputBufferBinding[i].Offset = offset;
+                    inputBufferBinding[i].SizeInBytes = arrayBufferView.byteLength;
+                    memcpy(
+                        uploadBuffer + offset,
+                        static_cast<int8_t*>(arrayBufferView.buffer) + arrayBufferView.byteOffset,
+                        arrayBufferView.byteLength);
+                    offset = offset + arrayBufferView.byteLength;
+                }
+            }
+        }
+        mUploadResource->Unmap(0, nullptr);
```

Comment:
OK, while I have a question, if we don't call unmap() , does it mean the CPU can always read or write this subresource? Will it be safe?

Reply:
Yes, the ID3D12Resource is always CPU-accessible when persistently mapped (= no Unmap is called). It is safe so long as the GPU does not read while the CPU is writing. This can be accomplished by the (correct) use of a GPU fence.

Reply:
We just need to call Map() when the resource gets created then Umap() once it is destroyed. If you don't call Unmap, nothing bad happens but D3D12 will complain.

Reply:
If so, I prefer keeping using unmap(), is that OK?

Reply:
We can still call Unmap() but ensure it gets called in a destructor and not each time you update it.

Reply:
So should here https://github.com/microsoft/DirectML/blob/master/Python/src/device.cpp#L505 also remove unmap()?